### PR TITLE
DOC/TST: Indexing with NA raises

### PR DESCRIPTION
**Pull request by TomAugspurger**
xref https://github.com/pandas-dev/pandas/issues/29556, https://github.com/pandas-dev/pandas/issues/28778

We're already doing the right thing on master. This just documents that behavior, and adds a handful of tests.

I'm not sure if there are existing tests I should be parameterizing. I only found a couple in `tests/indexing/`, which I parametrized over bool and boolean dtype. Will add more if I've missed any. In the meantime, I've written new tests.

### CODE REVIEW

File: doc/source/user_guide/boolean.rst
Comment (WillAyd) [inline doc/source/user_guide/boolean.rst]: Maybe use `pd.NA` instead of `None` just to reinforce other new features of the release?

File: pandas/core/arrays/boolean.py
Comment (WillAyd) [inline pandas/core/arrays/boolean.py]: Just out of curiosity in what way is this expensive? I would think should be a pretty quick op

File: pandas/tests/indexing/test_na_indexing.py
Comment (WillAyd) [inline pandas/tests/indexing/test_na_indexing.py]: Can we parametrize this? IIRC `nulls_fixture` wasn't appropriate but maybe need a `nulls_scalar_fixture` for these purposes

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/boolean.py
Comment (TomAugspurger) [reply pandas/core/arrays/boolean.py]: I suppose "relatively expensive" would be better. What I meant was that it's not a simple attribute lookup, since it makes two passes over the data.

If / when we have an optional mask, this will something like `self._mask is not None`.

File: pandas/core/arrays/boolean.py
Comment (jorisvandenbossche) [inline pandas/core/arrays/boolean.py]: we should probably not call it "has *nan*s" now .. :)

**Review by jorisvandenbossche — COMMENTED** [review]
This indeed seems to already work for Series etc, but not yet for the EAs itself:

```
In [22]: arr = pd.array([True, False, True], dtype="boolean") 

In [23]: arr[arr]  
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-23-9b15d846534b> in <module>
----> 1 arr[arr]

~/scipy/pandas/pandas/core/arrays/boolean.py in __getitem__(self, item)
    295                 return self.dtype.na_value
    296             return self._data[item]
--> 297         return type(self)(self._data[item], self._mask[item])
    298 
    299     def to_numpy(self, dtype=None, copy=False, na_value: "Scalar" = libmissing.NA):

IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices
```

So we need to add some check there that converts BooleanArray to numpy bool array if possible, or otherwise raise the informative error (we should already have some utility somewhere, since it is being done in the indexing machinery)

**Comment by TomAugspurger** [comment]
> but not yet for the EAs itself:

Good point. I think I'll add a base test for this, since all EAs should probably be able to handle it. Or is that too opinionated?

> (we should already have some utility somewhere, since it is being done in the indexing machinery)

It's `pandas.core.indexing.check_bool_indexer`, which converts the array-like to a boolean ndarray. I can make public somehow, if we want 3rd party EAs to handle these...

**Comment by jorisvandenbossche** [comment]
> I think I'll add a base test for this, since all EAs should probably be able to handle it. Or is that too opinionated?

It will probably give failures for external projects that implement EAs .. But maybe that is good, as indeed ideally they would handle this.

>  I can make public somehow, if we want 3rd party EAs to handle these...

That sounds as a good idea to me!

**Comment by TomAugspurger** [comment]
>>  I can make public somehow, if we want 3rd party EAs to handle these...

> That sounds as a good idea to me!

Do people have an initial preference on where to put this? I was thinking a new module in `pandas.api`, perhaps `pandas.api.indexing`. OTOH, this is mostly going to be used by EAs, and we already have `take` in `pandas.api.extensions`, so perhaps we put it with `take`?

**Comment by jorisvandenbossche** [comment]
I would put it like `take` api.extensions

File: pandas/tests/indexing/test_na_indexing.py
Comment (TomAugspurger) [reply pandas/tests/indexing/test_na_indexing.py]: What would be parametrized? The boolean array with missing values?

**Comment by TomAugspurger** [comment]
Added a couple base tests for `ExtensionArray.__getitem__(booleanarray)`, and updated our EAs to pass.

**Comment by TomAugspurger** [comment]
Fixed the mypy issues.

On performance, a BooleanArray takes about about 1.50x as long as an ndarray for a 10,000 element array

```python
In [5]: s = pd.Series(np.arange(10000))
   ...:
   ...: m1 = np.zeros(len(s), dtype="bool")
   ...: m2 = pd.array(m1, dtype="boolean")

In [6]: %timeit s[m1]
270 µs ± 10.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [7]: %timeit s[m2]
391 µs ± 30.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [8]: 391/270
Out[8]: 1.4481481481481482
```

I suspect this is do to the additional `.isna().any()` check we need for BooleanArray, though there may be an unnecessary copy.

Actually... we're doing *two* `isna().any()` checks, one in `is_bool_indexer` and one in `check_bool_array_indexer`. I didn't realize that `is_bool_indexer` also converted to an ndarray and checked for missing values. I'm guessing we'll want to combine those checks.

**Comment by TomAugspurger** [comment]
Pushed a perf / bugfix update, will see if it breaks things. I restructured `is_bool_indexer` to not do NA checking. But as I write this, I wonder why we even have `is_bool_indexer`? I'd rather just use `is_bool_dtype` and `check_bool_indexer`. Will look into that now.

Anyway, that turned up a bug in `DatetimeArray.__getitem__` which has been fixed and documented.

I'm probably going to stop looking at perf things now, but there's likely more to be done. We're still about 1.3x as long as ndarray, but only 8% of the time is spent in `check_array` doing the `isna().any()`, which should be the extent of the overhead. We're leaving a lot on the table I think: https://github.com/pandas-dev/pandas/issues/30349

File: pandas/tests/extension/base/getitem.py
Comment (jorisvandenbossche) [inline pandas/tests/extension/base/getitem.py]:
```suggestion
    def test_getitem_booleanarray_mask(self, data):
```

File: pandas/tests/extension/base/getitem.py
Comment (jorisvandenbossche) [inline pandas/tests/extension/base/getitem.py]:
```suggestion
    def test_getitem_booleanarray_mask_raises(self, data):
```

File: pandas/core/common.py
Comment (jorisvandenbossche) [inline pandas/core/common.py]: It was already there, but I find "vector" a strange word to use since we almost never that term in pandas context . Maybe "mask" would be better (to have a generic word for boolean array / series) ?

**Review by jorisvandenbossche — COMMENTED** [review]
This looks good to me

**Comment by TomAugspurger** [comment]
All green.

**Comment by TomAugspurger** [comment]
Planning to merge this later today if there aren't any objections.

File: pandas/api/extensions/__init__.py
Comment (jreback) [inline pandas/api/extensions/__init__.py]: @jbrockmendel ought to move is_bool_indexer to a more usual location, pandas.core.dtypes.common is better

File: pandas/core/arrays/boolean.py
Comment (jreback) [inline pandas/core/arrays/boolean.py]: this could easily be cached (and then updated on setitem / other mutation)

File: doc/source/reference/extensions.rst
Comment (jreback) [inline doc/source/reference/extensions.rst]: is there a real reason to expose check_book_array_indexer now? 
is_bool_indexer also shouldn't be in api.extension, api.types would be ok, though it was never meant to be public.

can we defer both of these as exposing to the public until we actually see use?

File: pandas/core/indexing.py
Comment (jreback) [inline pandas/core/indexing.py]: i would like to see these utilities split out of here (pandas.core.indexing is huge). but can certainly be done later.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/reference/extensions.rst
Comment (TomAugspurger) [reply doc/source/reference/extensions.rst]: Happy to defer exposing those utilities. Will push an update this evening.

**Comment by jorisvandenbossche** [comment]
Those utilities are useful for anybody implementing an ExtensionArray and who wants proper boolean indexing. Once this PR is merged, the geopandas tests will start failing, and to fix those, I can use those utilities.

**Comment by TomAugspurger** [comment]
I’m a bit concerned about the implementation of those. I think we can do the check and conversion with only one pass over the data (right now we do two). Adding them to the public api will possibly make changing them a bit harder. Not sure.

File: pandas/api/extensions/__init__.py
Comment (jbrockmendel) [reply pandas/api/extensions/__init__.py]: i think core.indexers

File: pandas/core/indexing.py
Comment (jbrockmendel) [inline pandas/core/indexing.py]: should we worry about shape here?  i.e. either handle or disallow ndim>1?

File: pandas/core/indexing.py
Comment (jbrockmendel) [reply pandas/core/indexing.py]: this is also simple enough it could go into core.indexers

File: pandas/core/indexing.py
Comment (TomAugspurger) [reply pandas/core/indexing.py]: I don't think so. This is just an extraction from `is_bool_indexer`, so on master we expect shape to be handled elsewhere. It's not clear to me that we'd want to always disallow ndim > 1

**Comment by TomAugspurger** [comment]
I've removed the indexing helpers from `api.extensions`. I think for now 3rd party EAs will be on their own. We can revisit that, but I think it's the right move for now.

**Comment by jbrockmendel** [comment]
(not a blocker) there's been a discussion about implementing an ExtensionIndex for Indexes backed by EAs.  what happens when we have an Index backed by IntegerArray with NAs?

**Comment by jreback** [comment]
> (not a blocker) there's been a discussion about implementing an ExtensionIndex for Indexes backed by EAs. what happens when we have an Index backed by IntegerArray with NAs?

not sure what you are asking.
this is exactly what an EAI would hold

**Comment by jbrockmendel** [comment]
> not sure what you are asking.

I think I saw how i got mixed up.  never mind.

**Comment by jorisvandenbossche** [comment]
I am personally -1 on releasing 1.0 without a helper function. IMO it would be a bit stupid to ask of each external EA implementor to implement this logic themselves.

**Comment by jorisvandenbossche** [comment]
Taking a step back: Tom, you say you are a bit concerned about the implementation, and the fact that it is now a two-step function (that needs to do potentially two passes over the values). Adding it to the public API might indeed make it more difficult to change this.
Do you have an idea how such a 1-pass API would look like? (we can also try to implement a public function right now still in terms of the two steps, but don't expose it as two separate steps; if we have an idea about the ideal API).

---

So the two steps right now are: 1) check if it is a boolean indexer and then 2) check the validity of the boolean indexer (the length, the nulls). 

So you get a pattern like this:

```
        ...
        elif com.is_bool_indexer(item):
            item = check_bool_array_indexer(self, item)
        ...
```

This can probably be done in a single function, although this single function would only need to raise if it is a boolean indexer, and otherwise pass through the non-boolean indexer (so personally I am not sure a single function would make this easier to read and follow the logic ..). 

Now what is the cost of the two functions? The two passes through the data, where does this happen? I am actually not fully sure: currently, we already check for missing values in `is_bool_indexer` and not in `check_bool_array_indexer`, although in  `check_bool_array_indexer` we do it implicitly again in `asarray(.., dtype=bool)`, as this will raise if there are missing values. 
Can we skip the null checking in `is_bool_indexer` and only do this in `check_bool_array_indexer` ?

**Comment by TomAugspurger** [comment]
> Can we skip the null checking in is_bool_indexer and only do this in check_bool_array_indexer ?

That’s my hope. I don’t know how much work that is off hand. But right now we’re doing the NA checking twice when we should only need to do it once. Will spend some time today or tomorrow on it.

**Comment by TomAugspurger** [comment]
So I think it’ll still be two functions, one to check the dtype, and a second to check for NA and convert to ndarray.

**Comment by jorisvandenbossche** [comment]
The reason that the dtype check is complicated, and not simply the already exposed `is_bool_dtype`, is because we also want to allow object dtype arrays with True/False values, right? 
We should maybe consider deprecating that in the future (boolean indexing with object dtype indexer), so as external EA author, I don't mind not implementing support for object dtype. In such a case, I could use `pd.api.types.is_bool_dtype` for the type check, which would avoid the need to expose a new `is_bool_indexer`, only needing the `check_bool_array_indexer`

**Comment by TomAugspurger** [comment]
@jorisvandenbossche can you take a look at https://github.com/pandas-dev/pandas/pull/30308/commits/21fd589d5003578340314e884b4cdcbbb0d6b71d? That's updated `DecimalArray.__getitem__` to handle boolean masks, and I think is representative of what 3rd party arrays would need to do.

If we're OK with that, I'll add `check_bool_array_indexer` to the public API. I'm not worried about its implementation changing. I'm still not sure where it should go though, but probably `api.extensions`.

**Comment by jreback** [comment]
we have api.indexers now -
seems like a good place

**Comment by TomAugspurger** [comment]
Yeah, just saw that. It seems reasonable.

**Comment by TomAugspurger** [comment]
All green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
very nice @TomAugspurger 

lots of testing!

File: pandas/core/indexers.py
Comment (jorisvandenbossche) [inline pandas/core/indexers.py]: this does not exist now (will do a PR)

File: pandas/core/indexers.py
Comment (jorisvandenbossche) [inline pandas/core/indexers.py]: Should we try to improve this error message? We know this is called in terms of indexing, and then something like "Cannot do boolean indexing with missing values, use fillna(True/False) ..." would be a much more useful error message than the message about conversion to numpy array.

**Review by jorisvandenbossche — COMMENTED** [review]
@TomAugspurger thanks, that function looks good!

On the location of `pandas.api.indexers`, no strong opinion on putting it there in se, but I am not sure this new function necessarily fits together with a rolling-window functionality (but then I also find `api.indexers` a strange name for rolling window functionality that has nothing to do with indexing, will comment about that elsewhere)

File: pandas/core/indexers.py
Comment (jorisvandenbossche) [reply pandas/core/indexers.py]: https://github.com/pandas-dev/pandas/pull/30725

**Comment by jorisvandenbossche** [comment]
While implementing support for this in GeoPandas, I ran into an issue with your example implementation for DecimalArray, basically being:

```
        def __getitem__(self, item):
            ....
            # array, slice.
            if pd.api.types.is_list_like(item):
                if not pd.api.types.is_array_like(item):
                    item = pd.array(item)
                dtype = item.dtype
                if pd.api.types.is_bool_dtype(dtype):
                    item = check_bool_array_indexer(self, item)
            return type(self)(self._data[item])
```

The problem with the above is that this also converts integers to IntegerArray, which is then not supported in the numpy indexing call (not fully sure why this doesn't come up in the decimal array's tests). 
Opened https://github.com/pandas-dev/pandas/issues/30738 for this about indexing with IntegerArray in general.

That issue is maybe not a blocker for 1.0 (it also doesn't work on 0.25), but we should think a moment if this might change how we want to expose an array indexer check function.

**Comment by jorisvandenbossche** [comment]
> not fully sure why this doesn't come up in the decimal array's tests).

Because you added a fix for IntegerArray ;)

File: pandas/core/arrays/sparse/array.py
Comment (jbrockmendel) [inline pandas/core/arrays/sparse/array.py:773]: why is this check_bool_indexer when in all the others its check_bool_array_indexer?  found this bc mypy is complaining about that the first arg should be an Index

**Pull request merged**
